File: elixir/bob/lib/bob.ex
defmodule Bob do
  def hey(input) do
    cond do
      is_silence?(input) ->
        "Fine. Be that way!"

      is_yell?(input) && String.ends_with?(String.trim(input), ["?"]) == false ->
        "Whoa, chill out!"

      is_yell?(input) == false && String.ends_with?(String.trim(input), ["?"]) ->
        "Sure."

      is_yell?(input) && String.ends_with?(String.trim(input), ["?"]) ->
        "Calm down, I know what I'm doing!"

      true ->
        "Whatever."
    end
  end

  defp is_silence?(input) do
    input |> String.trim() |> String.length() == 0
  end

  defp is_yell?(input) do
    result =
      input
      |> String.split()
      |> Enum.filter(fn w -> String.match?(w, ~r/[^0-9]\w+/) end)
      |> Enum.map(fn w -> String.match?(w, ~r{\w+}) && String.upcase(w) == w end)

    Enum.all?(result) && Enum.count(result) != 0
  end
end

File: fsharp/bob/Bob.fs
module rec Bob

open System.Text.RegularExpressions

let response (input: string) : string =
    match input with
    | Question -> "Sure."
    | Silence -> "Fine. Be that way!"
    | Yell -> "Whoa, chill out!"
    | YelledQuestion -> "Calm down, I know what I'm doing!"
    | Whatever -> "Whatever."


let (|Question|Silence|Yell|YelledQuestion|Whatever|) (input: string) =
    let trimmed = input.Trim()

    let noise =
        Regex.Replace(input, "[^a-zA-Z0-9 -?]", "").Trim()

    if noise.Length = 0 then
        Silence
    else if isAllUpper trimmed && trimmed.EndsWith("?") then
        YelledQuestion
    else if trimmed.EndsWith("?") then
        Question
    else if isAllUpper trimmed then
        Yell
    else
        Whatever

let isAllUpper input =
    let textOnly =
        Regex.Replace(input, "[^a-zA-Z -]", "").Trim()

    let upped = textOnly.ToUpper()
    textOnly.Length > 0 && textOnly = upped
